### ISSUE TITLE
Support recursion and distinct selection

### ISSUE THREAD

**Issue by jhalborg** [prisma/prisma#971]
Hey 👋

We have a use case that I've presented to @divyenduz  and @pantharshit00 , and have been asked to share it here for further discussion. 

_____

**The use case**
Essentially, it's a lot of teams of teams with users in each. The teams will have a hierarchy to them, often times defined by geographical relations. For example,

- A team, Oxford Street, has a list of users.
- Another team, London, has a list of users, as well as the Oxford Street and a third team, "The Big Ben Squad" in it.
- Yet another team, UK, has the London team, and, by parent ownership, the street teams as well.

**The requirements**
- A team is a collection of users
- Teams have a hierarchy to them, and teams higher up automatically include all users of child teams.
- Users can also be connected to teams higher up in the hierarchy without being included in the child teams. I.e., a user can be connected to "UK" without also being a part of neither "London" nor "Oxford Street"
- Users can be part of multiple teams in multiple levels of the hierarchy, for example both "London", "Oxford Street" and "UK"

As far as we can tell, it will require self-referencing models, which is supported by native MySQL using CTEs, but we're not sure if Photon will be able to handle it in a performant way.

**A pseudo-schema**
```
type User {
  partOfTeams: [Team]
}

type Team {
  users: [User!]
  parentTeam: Team
  childTeams: [Team]
}
```

The schema does not completely encapsulate the requirements mentioned in text - specifically that teams higher in the hierarchy automatically should include users from teams lower in the hierarchy

**Performance considerations**
We're afraid that if we want to get, say, all users connected to one of the top level hierarchies, say a country, that the query could potentially be very slow due to the recursive nature of the data

For example:

UK is at the top of a hierarchy. It has 2000 users connected directly to it. Below it are 80 major cities, each with an average of 1000 users connected directly to it. Below each of those is an average of 20 teams, each with an average of 200 users

When we want to get "How many users are in teams in the UK?", we should get

`2000 + (80*1000) + (20*200) = 86000` users 

We might not ever want to get a list of that magnitude unpaginated, but we'd want to get aggregates for each level in the hierarchy, and also be able to get paginated results of the tree on different entry points (i.e. users in UK, users in London ...)

**Skip duplicates**
Furthermore, we would like to be able to skip duplicates if a specific user is part of multiple teams - for example if a user is part of two teams in the same city, he should only count for 1 in the total of "number of users in UK", and only show once on that list, but should still show once for each of the specific teams

**Example**
Let's say we have four layers of teams for a tree

1. UK
2. London
3. London West, London East
4. Team A In London West, Team B in London West

Our user, Waldo, is part of both Team A and Team B. Another user, John, is part of the London team. 

If we want a user list that includes all users in the UK, we expect both John and Waldo to be returned. Ideally, Waldo would only be included once, in spite of being part of two teams.

**Proposed API**
How would one contruct such a query using Photon?

```
const teamsInTree = await photon.teams.findMany({
    where: { ?? } // not sure what a recursion API would look like, it needs a break condition
})
const users = await photon.users.findMany({
    where: {
      partOfTeam: teamsInTree.map((team) => team.id),
    },
    {
        skipDuplicates: true // any plans for something like this?
    }
  })
```

**Comment by pantharshit00** [comment]
I went through the photon spec and we do have an API laid out for Distinct selection: https://github.com/prisma/specs/tree/master/photonjs#distinct

It is not implemented yet.

**Comment by JensMadsen** [comment]
> I went through the photon spec and we do have an API laid out for Distinct selection: https://github.com/prisma/specs/tree/master/photonjs#distinct
> 
> It is not implemented yet.

link broken. select on distinct like feature would be very velcome. Hasura has it :-)

**Comment by samburgers** [comment]
Would be keen for an update on this too!

**Comment by pantharshit00** [comment]
It can be found here: https://github.com/prisma/specs/issues/356

**Comment by beeplin** [comment]
+1. `distinct` is necessary in most real world projects.

**Comment by thebiglabasky** [comment]
We implemented `distinct` in version 2.3.0 so I will close this issue.
If there is another more specific part which isn't covered after trying out `distinct`, happy to re-open or create a new one tailored to that use case.
Thanks!
